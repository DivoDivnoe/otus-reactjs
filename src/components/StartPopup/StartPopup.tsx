import React, { Component, ReactNode } from 'react';
import styled from '@emotion/styled';
import Button from '@/components/Button/Button';

export interface PopupProps {
  isVisible: boolean;
  submitHandler: (name: string) => void;
}

export interface PopupState {
  name: string;
}

const PopupWrapper = styled.div`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: rgba(8, 0, 18, 0.85);
`;

const PopupItem = styled.div`
  position: absolute;
  left: 50%;
  top: 50%;
  padding: 30px;
  border-radius: 7px;
  background-color: #021a60;
  transform: translate(-50%, -50%);
`;

const FormItem = styled.form`
  display: flex;
  flex-direction: column;
  width: 240px;
`;

const InputItem = styled.input`
  padding: 6px 10px 4px 10px;
  margin-bottom: 20px;
  font-size: 16px;
  line-height: 19px;
  color: #ffffff;
  border: 2px solid #058aba;
  border-radius: 3px;
  background-color: transparent;
  box-shadow: inset 1px 1px 3px rgb(0 0 0 / 30%), 1px 1px 0 rgb(0 0 0 / 10%);
`;

class StartPopup extends Component<PopupProps, PopupState> {
  state = {
    name: '',
  };

  render(): ReactNode {
    const { isVisible } = this.props;

    return (
      isVisible && (
        <PopupWrapper>
          <PopupItem>
            <FormItem data-testid='start-form' onSubmit={this._onSubmit}>
              <InputItem
                data-testid='name-input'
                value={this.state.name}
                placeholder={'Enter Your Name:'}
                onChange={this._onChange}
                required
              />
              <Button>Start</Button>
            </FormItem>
          </PopupItem>
        </PopupWrapper>
      )
    );
  }

  _onSubmit = (evt: React.FormEvent): void => {
    evt.preventDefault();

    this.props.submitHandler(this.state.name);
  };

  _onChange = (evt: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({ name: evt.target.value.trim() });
  };
}

export default StartPopup;
