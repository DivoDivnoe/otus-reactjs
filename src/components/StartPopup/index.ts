export { StartPopup, StartPopupProps } from './StartPopup';
