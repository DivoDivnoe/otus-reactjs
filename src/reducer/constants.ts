export enum ActionType {
  SET_USER = 'SET_USER',
  RESET_USER = 'RESET_USER',
  START_PLAYING = 'START_PLAYING',
  STOP_PLAYING = 'STOP_PLAYING',
  SET_SIZE = 'SET_SIZE',
  SET_SPEED = 'SET_SPEED',
  SET_FILL = 'SET_FILL',
  SET_MODEL = 'SET_MODEL',
  RESET_MODEL = 'RESET_MODEL',
}
