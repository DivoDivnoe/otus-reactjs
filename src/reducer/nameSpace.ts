enum NameSpace {
  GAME = 'game',
  USER = 'user',
}

export default NameSpace;
