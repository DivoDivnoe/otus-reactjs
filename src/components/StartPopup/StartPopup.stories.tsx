import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { StartPopup } from './';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer from '@/reducer';

export default {
  title: 'Example/StartPopup',
  component: StartPopup,
  argTypes: {
    signin: { action: 'sign' },
    setName: { action: 'setName' },
    resetName: { action: 'resetName' },
  },
} as ComponentMeta<typeof StartPopup>;

const store = createStore(reducer);

const Template: ComponentStory<typeof StartPopup> = (args) => (
  <Provider store={store}>
    <StartPopup {...args} />
  </Provider>
);

export const MainStartPopup = Template.bind({});
MainStartPopup.args = {
  currentName: 'Andrey',
};
