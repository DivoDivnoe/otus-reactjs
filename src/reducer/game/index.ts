import { combineReducers } from 'redux';
import ModelReducer from './model';
import IsPlayingReducer from './isPlaying';
import SizeReducer from './size';
import SpeedReducer from './speed';
import FillReducer from './fill';
import { Model } from '@/core';
import { BoardSize, SpeedType, FillType } from '@/constants';

export interface GameState {
  size: BoardSize;
  speed: SpeedType;
  fill: FillType;
  isPlaying: boolean;
  model: Model;
}

const reducer = combineReducers({
  size: SizeReducer,
  speed: SpeedReducer,
  fill: FillReducer,
  isPlaying: IsPlayingReducer,
  model: ModelReducer,
});

export default reducer;
