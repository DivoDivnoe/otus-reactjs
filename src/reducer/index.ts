import reducer from './reducer';

export { Operation, State } from './reducer';
export default reducer;
