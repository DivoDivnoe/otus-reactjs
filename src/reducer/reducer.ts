import { combineReducers } from 'redux';
import NameSpace from './nameSpace';
import gameReducer, { GameState } from '@/reducer/game';
import { APP_KEY } from './constants';
import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';
import userReducer, {
  ActionCreator as UserActionCreator,
  UserState,
  getUser,
} from '@/reducer/user';
import {
  ActionCreator as ModelActionCreator,
  getModel,
} from '@/reducer/game/model';
import {
  ActionCreator as SizeActionCreator,
  getSize,
} from '@/reducer/game/size';
import {
  ActionCreator as SpeedActionCreator,
  getSpeed,
} from '@/reducer/game/speed';

import {
  ActionCreator as FillActionCreator,
  getFill,
} from '@/reducer/game/fill';

import {
  ActionCreator as IsPlayingActionCreator,
  getIsPlaying,
} from '@/reducer/game/isPlaying';

export interface State {
  [NameSpace.GAME]: GameState;
  [NameSpace.USER]: UserState;
}

export type ThunkResult<R> = ThunkAction<R, State, undefined, AnyAction>;

export const Operation = {
  getStateFromLocalStorage: (): ThunkResult<void> => {
    return (dispatch) => {
      const rawState = window.localStorage.getItem(APP_KEY);

      if (rawState) {
        const state: State = JSON.parse(rawState);

        dispatch(UserActionCreator.setUser(getUser(state)));
        dispatch(SizeActionCreator.setSize(getSize(state)));
        dispatch(SpeedActionCreator.setSpeed(getSpeed(state)));
        dispatch(FillActionCreator.setFill(getFill(state)));
        dispatch(ModelActionCreator.setModel(getModel(state)));
        dispatch(IsPlayingActionCreator.setPlaying(getIsPlaying(state)));
      }
    };
  },
  saveStateToLocalStorage: (): ThunkResult<void> => {
    return (_dispatch, getState) => {
      const serializedState = JSON.stringify(getState());

      window.localStorage.setItem(APP_KEY, serializedState);
    };
  },
};

const reducer = combineReducers({
  [NameSpace.GAME]: gameReducer,
  [NameSpace.USER]: userReducer,
});

export default reducer;
