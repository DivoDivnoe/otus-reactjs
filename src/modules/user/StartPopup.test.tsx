import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { StartPopup } from './StartPopup';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer from '@/reducer';

describe('StartPopup', () => {
  it('is rendered correctly', () => {
    const store = createStore(reducer);

    const { asFragment } = render(
      <Provider store={store}>
        <StartPopup />
      </Provider>
    );
    expect(screen.getByRole('button', { name: 'Start' })).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it('input is empty', () => {
    const store = createStore(reducer);

    render(
      <Provider store={store}>
        <StartPopup />
      </Provider>
    );
    expect(screen.getByTestId('name-input')).toHaveValue('');
  });

  it('handles change event correctly', () => {
    const store = createStore(reducer);

    render(
      <Provider store={store}>
        <StartPopup />
      </Provider>
    );
    userEvent.type(screen.getByTestId('name-input'), 'Andrey');
    expect(screen.getByTestId('name-input')).toHaveValue('Andrey');
  });

  it('clears input after submit', () => {
    const store = createStore(reducer);

    render(
      <Provider store={store}>
        <StartPopup />
      </Provider>
    );
    userEvent.type(screen.getByTestId('name-input'), 'Andrey');
    fireEvent.submit(screen.getByTestId('start-form'));
    expect(screen.getByTestId('name-input')).toHaveValue('');
  });
});
