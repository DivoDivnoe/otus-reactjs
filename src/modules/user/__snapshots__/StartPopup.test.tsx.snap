// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartPopup is rendered correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1fng7gb"
    data-testid="start-popup"
  >
    <div
      class="css-1p0rybr"
    >
      <form
        class="css-qztj7d"
        data-testid="start-form"
      >
        <input
          class="css-1oh381d"
          data-testid="name-input"
          placeholder="Enter Your Name:"
          required=""
          value=""
        />
        <button
          class="css-som8ht"
        >
          Start
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
